app/com/arpnetworking/metrics/portal/reports/ReportExecution.java

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright 2019 Dropbox, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.arpnetworking.metrics.portal.reports;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableMultimap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.ImmutableSetMultimap;
+import com.google.common.collect.Maps;
+import com.google.inject.Injector;
+import com.google.inject.Key;
+import com.google.inject.name.Names;
+import com.typesafe.config.ConfigException;
+import models.internal.impl.DefaultReportResult;
+import models.internal.reports.Recipient;
+import models.internal.reports.Report;
+import models.internal.reports.ReportFormat;
+import models.internal.reports.ReportSource;
+
+import java.time.Instant;
+import java.util.Collection;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.stream.Collectors;
+
+/**
+ * Utilities for execution of {@link Report}s.
+ *
+ * @author Spencer Pearson (spencerpearson at dropbox dot com)
+ */
+public final class ReportExecution {
+
+    /**
+     * Render and send a report.
+     *
+     * Renders the report into all necessary formats (using the given Injector to look up the necessary {@link Renderer}s)
+     * and then sends each recipient their requested formats of {@link RenderedReport} (using the Injector to look up the {@link Sender}s).
+     *
+     * @param report The report to render and execute.
+     * @param injector A Guice injector to pull dependencies out of.
+     * @param scheduled The instant the report was scheduled for.
+     * @return A CompletionStage that completes when sending has completed for every recipient.
+     * @throws IllegalArgumentException if any necessary class is missing from the given Injector.
+     */
+    public static CompletionStage<Report.Result> execute(final Report report, final Injector injector, final Instant scheduled) {
+        try {
+            verifyDependencies(report, injector);
+        } catch (final ConfigException exception) {
```

Comment:
The documentation below indicates that an `IllegalArgumentException` is thrown.

As an aside, another way to do this would be to chain all these calls together (then the exception type doesn't matter).